### PR TITLE
Feat: provide filtering for OAuth error listener

### LINKED ISSUE #18575: Feat: provide filtering to OAuth error listener
## Description
The purpose is the try to provide custom filter  to [OAuthErrorEvent](https://github.com/abpframework/abp/blob/c742c099b3599fefb6756e83745faf4df3857087/npm/ng-packs/packages/oauth/src/lib/strategies/auth-flow-strategy.ts#L117) listener

## Todo
- [x] Create abstraction to `core` package
- [x] Implement to `oauth` package
- [x] Create documentation for `AuthErrorFilterService`
- [x] #18595

### PR DESCRIPTION
**Pull request by masum-ulu**
Resolves #18575

### Description

- Create an abstraction for filtering oauth error listener
- Listener clear storage but for some case you may want to prevent this behavior

### Checklist

- [ ] I fully tested it as developer / designer and created unit / integration tests
- [x] I documented it (or no need to document or I will create a separate documentation issue)

### CODE REVIEW

File: npm/ng-packs/packages/core/src/lib/abstracts/auth-error-filter.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  abstract remove(id: string): void;
+  abstract run(event: E): boolean;
+}
+
+export class AuthErrorFilterService<
+  T = AuthErrorEvent,
+  E = AuthErrorFilter,
+> extends AbstractAuthErrorFilter<T, E> {
+  get(id: string): T {
+    throw new Error('Import AbpOAuthModule from  @abp/ng.oauth or custom implementation');
```
Comment (mahmut-gundogdu) [inline npm/ng-packs/packages/core/src/lib/abstracts/auth-error-filter.ts]: I would move  the loginc a function that return the error.

File: npm/ng-packs/packages/core/src/lib/abstracts/auth-error-filter.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import { signal } from '@angular/core';
+import { AuthErrorEvent, AuthErrorFilter } from '../models';
+
+export abstract class AbstractAuthErrorFilter<T, E> {
+  readonly #filters = signal<Array<T>>([]);
+  filters = this.#filters.asReadonly();
```
Comment (mahmut-gundogdu) [inline npm/ng-packs/packages/core/src/lib/abstracts/auth-error-filter.ts]: We just define the readonly Signal. We shouldn't define private object in abstract

File: npm/ng-packs/packages/core/src/lib/models/auth-events.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  | 'session_terminated'
+  | 'session_unchanged'
+  | 'logout'
+  | 'popup_closed'
+  | 'popup_blocked'
+  | 'token_revoke_error';
+
+export abstract class AuthEvent {
+  readonly type: EventType;
+  constructor(type: EventType) {
```
Comment (mahmut-gundogdu) [inline npm/ng-packs/packages/core/src/lib/models/auth-events.ts]: we may define like constructor(public type: EventType)

File: npm/ng-packs/packages/oauth/src/lib/oauth.module.ts [hunk 4]
```diff
@@ -60,6 +61,7 @@ export class AbpOAuthModule {
         },
         OAuthModule.forRoot().providers as Provider[],
         { provide: OAuthStorage, useClass: AbpLocalStorageService },
+        { provide: AuthErrorFilterService, useExisting: OAuthErrorFilterService },
```
Comment (mahmut-gundogdu) [inline npm/ng-packs/packages/oauth/src/lib/oauth.module.ts:64]: why useExistsing? why not useClass?

**Review by mahmut-gundogdu — CHANGES_REQUESTED** [review]

File: npm/ng-packs/packages/oauth/src/lib/oauth.module.ts (hunk 4 above)
Comment (masum-ulu) [reply npm/ng-packs/packages/oauth/src/lib/oauth.module.ts:64]: ProvidedIn root

image

**Review by mahmut-gundogdu — APPROVED** [review]

**Pull request merged**